lmdeploy/vl/templates.py

```diff
@@ -428,6 +428,73 @@ class GLM4VChatTemplateWrapper(VLChatTemplateWrapper):
     pass
 
 
+class MolmoChatTemplateWrapper(VLChatTemplateWrapper):
+
+    async def async_collect_pil_images(
+            self, messages: List[Dict]) -> List[Tuple[PIL.Image.Image, Dict]]:
+        """collect images from messages.
+
+        Args:
+            messages (List[Dict]): a user request of GPT4V message format
+        """
+        if isinstance(messages, Dict):
+            messages = [messages]
+        assert isinstance(messages, List)
+
+        out_messages = []
+
+        def _inner_call(i, in_messages, out_messages):
+            role = in_messages[i]['role']
+            content = in_messages[i]['content']
+            if role != 'user' or isinstance(content, str):
+                # means message is user's prompt input or assistant's prompt,
+                # returning it directory
+                out_messages.append(in_messages[i])
+                return
+            # the role is a user and the content is a list
+            assert isinstance(content, List)
+            message = dict(role=role, content='', images=[])
+            for item in content:
+                # 'image_url': means url or local path to image.
+                # 'image_data': means PIL.Image.Image object.
+                if item['type'] == 'image_url':
+                    try:
+                        image = load_image(item['image_url']['url'])
+                        message['images'].append(image)
+                    except KeyError:
+                        logger.error(f'invalid format {message}')
+                elif item['type'] == 'image_data':
+                    try:
+                        image = load_image(item['image_data']['data'])
+                        message['images'].append(image)
+                    except KeyError:
+                        logger.error(f'invalid format {message}')
+                elif item['type'] == 'text':
+                    message['content'] = item['text']
+                else:
+                    logger.error(f'unexpected content type {message}')
+            out_messages.append(message)
+
+        await asyncio.gather(*[
+            asyncio.get_event_loop().run_in_executor(None, _inner_call, i,
+                                                     messages, out_messages)
+            for i in range(len(messages))
+        ])
+        #
```

Comment:
Concurrency cannot be used here or the input of `[user, assistant, user]` may become `[assistant, user, user]`

Reply:
And after fixing the order, I find the position of `Assistant` seems not right. I decode the input token and the result is 
```
" User: Describe the two images in detail. Assistant:Assistant: xxx...User: What are the similarities and differences between these two images. "  
```
(use the multi-round example [here](https://github.com/InternLM/lmdeploy/blob/main/docs/en/multi_modal/qwen2_vl.md))


If the input only has one user content and decode the input ids, the result is 
```
" User: Describe the two images in detail. Assistant:"
```

Reply:
Fixed